# Homework Introduction to Python

$\textbf{Ubah nama di bawah yaa!}$

- Nama: Almas Rausan Fikri
- IPK / Nilai Rapot: Nilai Rapot

In [1]:
import numpy as np

# please import as much as you need

# Normal

## 1. Operasi Variabel

> Misal kita memiliki toko sepatu. Modal membeli 1 kodi sepatu adalah Rp 1.000.000. Setiap sepatu dijual seharga Rp 80.000. Buatlah variabel modal, kuantitas, dan harga_jual sesuai informasi di atas. Hitung profit yang didapatkan jika semua sepatu berhasil terjual, dan masukkan ke variabel profit.

Saran langkah-langkah pengerjaan
1. Convert terlebih dahulu kodi dalam jumlah barang (1 kodi = xyz buah sepatu)
2. Gunakan operasi matematika sederhana untuk menghitung profit menggunakan tiga buah variabel yang sudah ditentukan

In [2]:
# 1 kodi = 20 buah
modal = 1000000
kuantitas = 20
harga_jual = 80000

revenue = kuantitas * harga_jual
print('revenue =', revenue)

profit = revenue - modal
print('profit =', profit)

revenue = 1600000
profit = 600000


Jadi, dari perhitungan di atas, dapat disimpulkan bahwa apabila semua sepatu bisa terjual, total sales adalah sebesar Rp 1.600.000, dengan modal Rp 1.000.000, maka toko sepatu kita untung Rp 600.000

## 2. Package Python dan Dictionary

> Kita akan coba mengakses data scientific article dari PLOS corpus
    Menggunakan API request dengan kode berikut : 
```
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
```
**perhatikan tanda kutipnya tidak salah ketika meng-copy ke dalam python

Pertanyaannya : 
Convert response.content dari request tersebut dari JSON Format ke dalam Dictionary, dan print-kan key di dalamnya apa saja

Saran langkah-langkah pengerjaan
1. Ikuti code dari soal, kemudian ekstrak value dari variable response menggunakan dot (.) content menjadi sebuah variabel baru (dictionary)
2. Gunakan library json, convert hasil content menggunakan json.loads
3. Print keys dari dictionary tersebut


In [3]:
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')

import json
dataset_python = json.loads(response.content)

print(dataset_python.keys())
print(dataset_python['response'].keys())
print(dataset_python['response']['docs'][0].keys())

dict_keys(['response'])
dict_keys(['numFound', 'start', 'docs'])
dict_keys(['id', 'abstract'])


Jadi, dari url link di atas, kami menyimpulkan bahwa response dari API plos.org memiliki hierarki sebagai berikut:

1. Raw dataset berisi dictionary 'response'
2. Dictionary response berisi keys seperti 'numFound', 'start', dan 'docs'. numFound berisi integer yang menunjukkan berapa jumlah jurnal yang ditemukan, sedangkan isi dari jurnal terdapat pada key 'docs'
3. key 'docs' berisikan list, masing-masing list berisi 'id' dan 'abstract' dari jurnal yang ditemukan


# Intermediate

## 3. Operasi dalam List

> Buatlah list dengan 10 nama kota di indonesia, kemudian:
1. Urutkan nama kota di dalam list berdasarkan abjad (a)
2. Ambil 4 kota pertama dari list tersebut (yang telah diurutkan) (b)
3. Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan) (c)

Saran langkah-langkah pengerjaan
1. Tidak ada ketentuan nama kota (misal berdasarkan provinsi X atau pulau Y), dibebaskan untuk memilih. Buat nama-nama kota tersebut dalam sebuah list.
2. Gunakan fungsi sorted(X) untuk mengurutkan secara kecil-ke-besar (sesuai abjad), dimana X adalah contoh variabel list pada langkah (a)
3. Gunakan hasil poin (a), urutkan nama kota di dalam list berdasarkan abjad, untuk mengerjakan poin (b) dan (c). 

In [4]:
kota_raw = ['Sabang', 'Binjai', 'Tebing Tinggi', 'Siak Sri Indrapura', 'Pringsewu', 
            'Depok', 'Pati', 'Banjarnegara', 'Banyumas', 'Merauke']

kota_sorted = sorted(kota_raw)
empat_pertama = kota_sorted[:4]
tiga_terakhir = kota_sorted[-3:]

print(kota_sorted)
print(empat_pertama)
print(tiga_terakhir)

['Banjarnegara', 'Banyumas', 'Binjai', 'Depok', 'Merauke', 'Pati', 'Pringsewu', 'Sabang', 'Siak Sri Indrapura', 'Tebing Tinggi']
['Banjarnegara', 'Banyumas', 'Binjai', 'Depok']
['Sabang', 'Siak Sri Indrapura', 'Tebing Tinggi']


Jadi, dari operasi list di atas, dapat disimpulkan bahwa dari 10 kota yang telah dipilih:
- Urutan kota secara abjad adalah: 'Banjarnegara', 'Banyumas', 'Binjai', 'Depok', 'Merauke', 'Pati', 'Pringsewu', 'Sabang', 'Siak Sri Indrapura', 'Tebing Tinggi'
- 'Banjarnegara', 'Banyumas', 'Binjai', 'Depok' menempati 4 kota urutan pertama secara abjad
- 'Sabang', 'Siak Sri Indrapura', 'Tebing Tinggi' menempati 3 kota urutan terakhir secara abjad

## 4. Function dalam List

> Team sales memiliki list total transaksi dari 10 merchant pada bulan Juni 2020, yaitu: 
```
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
```
1. Trx minimum
2. Rata-rata 
3. Trx maksimal

Saran langkah-langkah pengerjaan
1. Copy variabel total_transaksi yang ada di soal ke notebook kita (atau ketik ulang boleh)
2. Kemudian, hitung menggunakan function operasi dalam list yang telah dipelajari di kelas pada topik list. Print hasilnya setiap poin (a), (b), dan (c)

In [5]:
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]

print('1. Trx minimum =', min(total_transaksi))
print('2. Rata-rata =', sum(total_transaksi)/len(total_transaksi))
print('3. Trx maksimal =', max(total_transaksi))

1. Trx minimum = 10
2. Rata-rata = 25.6
3. Trx maksimal = 45


Jadi, dari hasil operasi list di atas, dapat disimpulkan bahwa nilai minimal transaksi sebesar 10, max sebesar 45, dan rata-rata sebesar 25.6

# Advanced

## 5. NumPy Array

> Buatlah array dari 5 bilangan ganjil pertama (dimulai dari 1), Kemudian lakukan perintah berikut : 
1. Kali-kan array tersebut dengan 10
2. Bagi-kan array (hasil dari bagian a) dengan 3
3. Kali-kan array (hasil dari bagian b) dengan array baru :
`[-45, 43, 3, 56, 5]`
4. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c) 

Saran langkah-langkah pengerjaan
1. HATI-HATI dalam pengerjaan, karena apabila poin sebelumnya salah, maka selanjutnya akan kebawa salah (misal (b) salah, maka poin (c) sampai (d) akan salah)
2. Untuk array 5 bilangan ganjil pertama boleh manual boleh menggunakan fungsi, dimulai dari 1
3. Ikuti perintah soal, gunakan numpy array, jangan list. Sesuaikan dengan permintaan soal pada setiap butir.



In [6]:
odd_list = [1,3,5,7,9]
boku_no_array = np.array(odd_list)

# 1
boku_no_array = boku_no_array * 10
print(boku_no_array)

# 2
boku_no_array = boku_no_array / 3
print(boku_no_array)

# 3
kimi_no_array = np.array([-45, 43, 3, 56, 5])
boku_no_array *= kimi_no_array
print(boku_no_array)

# 4
print("index angka terbesar =", boku_no_array.argmax())
print("index angka terkecil =", boku_no_array.argmin())

[10 30 50 70 90]
[ 3.33333333 10.         16.66666667 23.33333333 30.        ]
[-150.          430.           50.         1306.66666667  150.        ]
index angka terbesar = 3
index angka terkecil = 0


Jadi, dari hasil operasi numpy di atas, dapat disimpulkan bahwa setiap elemen dalam numpy array dapat dioperasikan seperti perkalian, pembagian, dan operasi dengan array lain.

Operasi numpy array dengan integer akan mengoperasikan setiap elemen array dengan bilangan tersebut. Sedangkan operasi dengan array lain akan dilakukan antar index yang sama 0-0, 1-1, dst.

Numpy array juga punya built-in method yaitu argmax() dan argmin() untuk mencari indeks dari nilai maksimal dan minimal dari array tersebut.

# Terimakasih, salam dari Banyumas